symbolic-ppdb/src/format/sourcelinks.rs

```diff
@@ -0,0 +1,201 @@
+            case-insensitively to documents and the resulting URL is used to download source. The document
+            may contain an asterisk to represent a wildcard in order to match anything in the asterisk's
+            location. The rules for the asterisk are as follows:
+                1. The only acceptable wildcard is one and only one '*', which if present will be replaced by a relative path.
+                2. If the file path does not contain a *, the URL cannot contain a * and if the file path contains a * the URL must contain a *.
+                3. If the file path contains a *, it must be the final character.
+                4. If the URL contains a *, it may be anywhere in the URL.
+            */
+            let key = doc.0.to_lowercase();
+            let url = doc.1.as_str().ok_or_else(Self::err)?.into();
+            let pattern = if key.ends_with('*') {
+                Pattern::Prefix(key[..(key.len() - 1)].into())
+            } else {
+                Pattern::Exact(key)
+            };
+            self.rules.push(Rule { pattern, url });
+            self.sorted = false;
+        }
+        Ok(())
+    }
+
+    fn err() -> FormatError {
+        FormatError {
+            kind: FormatErrorKind::InvalidSourceLinkJson,
+            source: None,
+        }
+    }
+
+    /// Sort internal rules. This must be called before [Self::resolve].
+    pub fn sort(&mut self) {
+        // Put Exact matches first, then sort by the Prefix length, longest to shortest.
+        self.rules.sort_unstable_by(|a, b| match &a.pattern {
+            Pattern::Exact(_) => Ordering::Less,
+            Pattern::Prefix(a) => match &b.pattern {
+                Pattern::Exact(_) => Ordering::Greater,
+                Pattern::Prefix(b) => b.len().cmp(&a.len()),
+            },
+        });
+        self.sorted = true;
+    }
+
+    /// Resolve the path to a URL.
+    pub fn resolve(&self, path: &str) -> Option<String> {
+        // Must be sorted first so we can return on the first match, which is guaranteed to be the most specific.
+        if !self.sorted {
+            return None;
+        }
+
+        // Note: this is currently quite simple, just pick the first match. If we needed to improve
+        // performance in the future because we encounter PDBs with too many items, we can do a
+        // prefix binary search, for example.
+        let path_lower = path.to_lowercase();
+        for rule in &self.rules {
+            match &rule.pattern {
+                Pattern::Exact(value) => {
+                    if value == &path_lower {
+                        return Some(rule.url.clone());
+                    }
+                }
+                Pattern::Prefix(value) => {
+                    if path_lower.starts_with(value) {
+                        let replacement = path
+                            .get(value.len()..)
+                            .map(|v| v.replace('\\', "/"))
+                            .unwrap_or_default();
```

Comment:
```suggestion
                    if let Some(rest) = path_lower.strip_prefix(value) {
                        let replacement = rest.replace('\\', "/");
```

Reply:
actually, I had to roll this back - it's not functionally the same - the original code preserves input case.

Reply:
Right, missed that, sorry. I think you could still do
```suggestion
                    if path_lower.starts_with(value) {
                        let replacement = path
                            .get(value.len()..).unwrap_or_default()
                            .replace('\\', "/");
```
to get rid of the `map`.